src/wcl/rope.h

```diff
@@ -58,6 +60,9 @@ class rope_impl_pair : public rope_impl_base {
     pair.first->write(ostream);
     pair.second->write(ostream);
   }
+  void drop_last() override {
```

Comment:
I think this would be just as efficient if it returned an std::shared_ptr instead. The method on the rope can still be a mutation but I think the impls should be immutable.

But also I'm wondering if this is really what's needed? Why do we want this method?

Reply:
API convenience mostly. There are several cases where you have loops of things with a common separator that shouldn't have a trailing instance of the separator

ie

`a <space> b <space> c <space>`

`drop_last` drops the last rope, which in every use case has been the last separator

We could avoid it but partly unrolling the loop but I didn't really want to write the code for that

Reply:
I see. Let's only use it when we have code like this then

```
for (...) {
...
  space(builder);
}
builder.drop_last();
```

or

```
for (...) {
...
  builder.emit(",");
}

builder.drop_last();
```

I think `drop_last` only works correctly right now on builder because builder concats using pair in a certain way. Maybe builder needs a stack of ropes and an "undo" instead?

We could also make a special function for doing these kinds of loops instead but you likely need a couple since the CST doesn't traverse as nicely as a vector.

Reply:
yeah, it's only being done like that so far. We can keep an eye out for it in review and kinda feel it out for a better API over time. I wasn't a big fan of it either but it seemed like the most sensical way to do it